[DominanceFrontier] make iterating dereferenced DominanceFrontierBase::find deterministic

wenju-he opened the pull request:
DominanceFrontierBase::find returns an iterator to a set of dominance
frontiers. When we dereference the iterator and then iterate over the
set, the order of the dominance frontiers is indeterministic.
This results in indeterministic IR output in our downstream use case
that depends on the order.

wenju-he committed:
[DominanceFrontier] make DominanceFrontierBase::find iterator deterministic

When we iterate over a basic block's dominance frontiers returned by
DominanceFrontierBase::find, the order of the dominance frontiers is
indeterministic. This results in indeterministic IR output in our
downstream use case that depends on the order.

wenju-he commented:
@fhahn could you please help to review? I don't have permission to add you as reviewer.

fhahn commented:
Could you clarify what the problem is exactly? The description says it makes the `DominanceFrontierBase::find` iterator deterministic, but `find` returns an iterator to the map, not the set you are changing?

So IIUC the patch makes iterating over the set of nodes for a frontier deterministic. Overall I agree that making this deterministic is desirable, I am a bit curious why this issue hasn't shown up before with current in tree users, and if there's some way we can test this in upstream.

wenju-he commented:
> The description says it makes the `DominanceFrontierBase::find` iterator deterministic, but `find` returns an iterator to the map, not the set you are changing?

Yes, `find` returns an iterator to the map. I meant that when I dereference the iterator and then iterator over the dereferenced value which is a set.
I've updated title and commit message.

> I am a bit curious why this issue hasn't shown up before with current in tree users

I find that the main use of DominanceFrontierBase is in class DominanceFrontier and MachineDominanceFrontier.

DominanceFrontierAnalysis is used in RegionInfo, the two following uses of `find` needn't to care about order:
https://github.com/llvm/llvm-project/blob/3651f377f632d9152f2bd8fc2f9302ec9f9bdd5e/llvm/include/llvm/Analysis/RegionInfoImpl.h#L571-L579
https://github.com/llvm/llvm-project/blob/3651f377f632d9152f2bd8fc2f9302ec9f9bdd5e/llvm/include/llvm/Analysis/RegionInfoImpl.h#L584-L594

I found two uses of `DomSetType` in MachineDominanceFrontier, and both of them only use set member functions and thus needn't to care about order:
https://github.com/llvm/llvm-project/blob/8291147356e5bc81d5bd3b904e760c689543dc9a/llvm/include/llvm/CodeGen/MachineDominanceFrontier.h#L77
https://github.com/llvm/llvm-project/blob/8291147356e5bc81d5bd3b904e760c689543dc9a/llvm/include/llvm/CodeGen/MachineDominanceFrontier.h#L93

> if there's some way we can test this in upstream.

This is a good suggestion. Unfortunately I don't have idea at the moment.

kuhar commented on llvm/include/llvm/Analysis/DominanceFrontier.h:
```diff
@@ -39,7 +39,7 @@ class raw_ostream;
 class DominanceFrontierBase {
 public:
-  using DomSetType = std::set<BlockT *>;                // Dom set for a bb
+  using DomSetType = SetVector<BlockT *>;               // Dom set for a bb
```
Can you add a comment explaining the choice of  `SetVector`?

wenju-he replied:
done

kuhar reviewed (commented):
Right, `set<ptr>` is as good as `unordered_set<ptr>`. Would be nice to have a test that expects a specific order regardless of pointer values.

wenju-he committed:
add a comment for SetVector

wenju-he committed:
check order is deterministic in test

wenju-he commented:
> Would be nice to have a test that expects a specific order regardless of pointer values.

Done. Updated the test. 
In the original test `new_pm_test.ll`, the print order is always deterministic w/ or w/o this PR.
I added `a` prefix to a few basic block labels so that the test will fail w/o this PR.

kuhar reviewed (approved):
LGTM. Because this is a piece of core infra, I'd give it another 24h before landing to see if anyone spots potential issues.

kuhar commented on llvm/include/llvm/Analysis/DominanceFrontier.h:
```diff
@@ -39,7 +39,9 @@ class raw_ostream;
+  // Dom set for a bb. Use SetVector to make iterating dom frontiers of a bb
+  // deterministic.
+  using DomSetType = SetVector<BlockT *>;
   using DomSetMapType = std::map<BlockT *, DomSetType>; // Dom set map
```
Separately from this PR, I wonder if we should use `DenseMap` here.

wenju-he replied:
https://llvm.org/docs/CodingStandards.html says `llvm::DenseMap should almost always be used instead of std::map`, shall I change it to DenseMap?

kuhar replied:
I don't see why not. But I'd open a separate PR for this.

wenju-he replied:
thank you all for the review.

> I don't see why not. But I'd open a separate PR for this.

opened a new pr: https://github.com/llvm/llvm-project/pull/70403

fhahn reviewed (approved):
LGTM, thanks!
